js/tabs.js

```diff
@@ -322,6 +309,54 @@ Tabs.prototype.openFiles = function() {
       this.openFileEntry.bind(this));
 };
 
+/**
+ * @param {function()} callback
+ * Invoke the save dialog for all tabs with unsaved progress. Does not close any tabs.
+ */
+Tabs.prototype.promptAllUnsaved = function(callback) {
+  this.promptAllUnsavedFromIndex_(0, callback);
+}
+
+Tabs.prototype.promptAllUnsavedFromIndex_ = function(i, callback) {
+  if (this.tabs_.length > i) {
+    var tab = this.tabs_[i];
+    if (tab.isSaved()) {
+      this.promptAllUnsavedFromIndex_(i + 1, callback);
+    } else {
+      this.showTab(this.tabs_[i].getId());
+      this.promptSave_(tab, function(answer) {
+        if (answer === 'yes') {
+          this.save(tab, false /* close */);
+        } else if (answer === 'cancel') {
+          return;
+        }
+        this.promptAllUnsavedFromIndex_(i + 1, callback);
+      }.bind(this));
+    }
+  }
+  else {
+    callback();
+  }
+}
+
+Tabs.prototype.promptSave_ = function(tab, callbackShowDialog) {
+  var filename = tab.getName();
+  if (filename.length > 25) {
+    filename = filename.slice(0, 22).concat('...');
+  }
+
+  this.dialogController_.setText(
```

Comment:
A side note --- this feels like something that a Builder pattern or some restructuring could help with. Not something for this patch, just an observation that this is a little difficult to read atm.

Reply:
Ack